src/workerd/api/sockets.c++

```diff
@@ -511,4 +513,117 @@ jsg::Ref<Socket> SocketsModule::connect(
+
+  kj::Promise<void> connect(kj::StringPtr host,
+      const kj::HttpHeaders& headers,
+      kj::AsyncIoStream& connection,
+      ConnectResponse& response,
+      kj::HttpConnectSettings settings) override {
+    JSG_FAIL_REQUIRE(TypeError,
+        "connect is not something that can be done on a fetcher converted from a socket");
+  }
+
+  kj::Promise<void> prewarm(kj::StringPtr url) override {
+    KJ_UNIMPLEMENTED("prewarm() not supported on StreamWorkerInterface");
+  }
+
+  kj::Promise<ScheduledResult> runScheduled(kj::Date scheduledTime, kj::StringPtr cron) override {
+    KJ_UNIMPLEMENTED("runScheduled() not supported on StreamWorkerInterface");
+  }
+
+  kj::Promise<AlarmResult> runAlarm(kj::Date scheduledTime, uint32_t retryCount) override {
+    KJ_UNIMPLEMENTED("runAlarm() not supported on StreamWorkerInterface");
+  }
+
+  kj::Promise<CustomEvent::Result> customEvent(kj::Own<CustomEvent> event) override {
+    return event->notSupported();
+  }
+
+ private:
+  kj::Own<kj::AsyncIoStream> stream;
+  const kj::HttpHeaderTable& headerTable;
+};
+
+// Implementation of the custom factory for creating WorkerInterface instances from a socket
+class StreamOutgoingFactory final: public Fetcher::OutgoingFactory {
+ public:
+  StreamOutgoingFactory(kj::Own<kj::AsyncIoStream> stream, const kj::HttpHeaderTable& headerTable)
+      : stream(kj::mv(stream)),
+        headerTable(headerTable) {}
+
+  kj::Own<WorkerInterface> newSingleUseClient(kj::Maybe<kj::String> cfStr) override {
+    JSG_ASSERT(stream.get() != nullptr, Error,
+        "Fetcher created from internalNewHttpClient can only be used once");
+    // Create a WorkerInterface that wraps the stream
+    return kj::heap<StreamWorkerInterface>(kj::mv(stream), headerTable);
+  }
+
+ private:
+  kj::Own<kj::AsyncIoStream> stream;
+  const kj::HttpHeaderTable& headerTable;
+};
+
+jsg::Promise<jsg::Ref<Fetcher>> SocketsModule::internalNewHttpClient(
+    jsg::Lock& js, jsg::Ref<Socket> socket) {
+
+  // TODO(soon) check for nothing to read, this will require things using a promise so this function
+  // must remain returning a jsg::Promise waiting on a TODO for releaseLock
+
+  // Flush the writable stream before taking the connection stream to ensure all data is written
+  // before the stream is detatched
+  return socket->getWritable()->flush(js).then(
+      js, JSG_VISITABLE_LAMBDA((socket = kj::mv(socket)), (socket), (jsg::Lock & js) mutable {
+        auto& ioctx = IoContext::current();
+
+        // Create our custom factory that will create client instances from this socket
+        kj::Own<Fetcher::OutgoingFactory> outgoingFactory = kj::heap<StreamOutgoingFactory>(
+            socket->takeConnectionStream(js), ioctx.getHeaderTable());
+
+        // Create a Fetcher that uses our custom factory
+        auto fetcher = js.alloc<Fetcher>(
+            ioctx.addObject(kj::mv(outgoingFactory)), Fetcher::RequiresHostAndProtocol::YES);
+
+        return kj::mv(fetcher);
+      }));
```

Comment:
@dom96 I think there is something wrong with flushing writes, can you take a look.

Reply:
Looks like you figured out the issue since the CI is passing now.

This code seems fine, but two things:

* Should the flushing logic actually live in `takeConnectionStream`?
* I'm not very familiar with how JSG_VISITABLE_LAMBDA works, maybe @jasnell can comment on whether this usage of it is correct. In particular I wonder if this should capture `JSG_THIS` in the same way that the `JSG_VISITABLE_LAMBDA` captures it in `startTls`.